ArduCopter/mode_auto.cpp

```diff
@@ -1096,8 +1096,33 @@ void Copter::ModeAuto::do_nav_wp(const AP_Mission::Mission_Command& cmd)
 
     // if no delay as well as not final waypoint set the waypoint as "fast"
     AP_Mission::Mission_Command temp_cmd;
+    bool fast_waypoint = false;
     if (loiter_time_max == 0 && mission.get_next_nav_cmd(cmd.index+1, temp_cmd)) {
-        copter.wp_nav->set_fast_waypoint(true);
+
+        // whether vehicle should stop at the target position depends upon the next command
+        switch (temp_cmd.id) {
+            case MAV_CMD_NAV_WAYPOINT:
+            case MAV_CMD_NAV_LOITER_UNLIM:
+            case MAV_CMD_NAV_LOITER_TURNS:
+            case MAV_CMD_NAV_LOITER_TIME:
+            case MAV_CMD_NAV_LAND:
+            case MAV_CMD_NAV_SPLINE_WAYPOINT:
+                // if next command's lat, lon is specified then do not slowdown at this waypoint
+                if ((temp_cmd.content.location.lat != 0) && (temp_cmd.content.location.lng != 0)) {
+                    fast_waypoint = true;
+                }
+                break;
+            case MAV_CMD_NAV_RETURN_TO_LAUNCH:
+                // do not stop for RTL
+                fast_waypoint = true;
```

Comment:
I wonder what happens if we are really close to home in this case.

Reply:
@OXINARF, thanks very much for the review.  I also wondered about what would happen if the final waypoint was very close to home (or the next waypoint's target was the same as the current waypoint) so I ran a test and the result is OK.

Below is a screen shot of the mission:
![rtl-test-mission](https://user-images.githubusercontent.com/1498098/53459286-8f893380-3a7c-11e9-9c25-4912bcbad642.png)

And here is the resulting roll and pitch angles.  They are totally under control - nothing like the drastic pitch back we saw before this PR's fix.
![rtl-test-attitude](https://user-images.githubusercontent.com/1498098/53459315-ab8cd500-3a7c-11e9-9275-da4f18e3ec20.png)

The reason it is OK is that even with a fast-waypoint, the vehicle can slowdown in time to stop at the target.  The bug was that the target point was being set closer than the command's target and LAND also uses the Loiter controller to slowdown which is more aggressive than the waypoint controller's slowdown.